File: ecom-backend/customer-service/src/main/java/devops/ecom/customerservice/service/ShoppingCartService.java
package devops.ecom.customerservice.service;

import devops.ecom.customerservice.dao.ShoppingCart;
import devops.ecom.customerservice.dao.ShoppingCartItem;
import devops.ecom.customerservice.exceptions.CustomerNotFoundException;
import devops.ecom.customerservice.model.AddItemRequest;

public interface ShoppingCartService {
    ShoppingCartItem createItem(AddItemRequest addItemRequest ) ;
    ShoppingCart addItemToCart(AddItemRequest addItemRequest) throws CustomerNotFoundException;
    ShoppingCart removeItemFromCart(String customerId , String productId) throws CustomerNotFoundException;
    ShoppingCart updateItemInCart(ShoppingCartItem item, AddItemRequest addItemRequest, ShoppingCart cart) ;
}

File: ecom-front/src/app/services/shoppingCartService/shopping-cart.service.ts
import { Injectable } from '@angular/core';
import {HttpClient} from "@angular/common/http";
import {Observable} from "rxjs";
import {AddItemRequest, ShoppingCart} from "../../models/ShoppingCart";
import {DeleteProductReq} from "../../models/common.model";
import {environment} from "../../../environments/environment";

@Injectable({
  providedIn: 'root'
})
export class ShoppingCartService {
  private customerService:string =environment.customerService
  constructor(private http : HttpClient) { }

  getShoppingCartOfCustomer(customerId: string):Observable<ShoppingCart> {
        return  this.http.get<ShoppingCart>(this.customerService + "/customers/"+ customerId +"/shoppingCart") ;
  }

  calcTotalPrice(cart :ShoppingCart):number {
    let totalPrice:number = 0;
    cart.items.forEach(
      value => totalPrice = totalPrice + (value.quantity * value.product.productPrice.price)
    )
    return totalPrice ;
  }

  addProductToShoppingCart(addProductReq: AddItemRequest):Observable<ShoppingCart> {
    return  this.http.post<ShoppingCart>(this.customerService + "/api" , addProductReq) ;
  }

  deleteItemFromCart(req: DeleteProductReq):Observable<ShoppingCart>{
    return  this.http.delete<ShoppingCart>(this.customerService + "/api/" +req.customerId+"/"+req.productId) ;
  }
}
